src/compartment.c

```diff
@@ -239,22 +245,79 @@ void
 comp_register_ddc(struct Compartment* new_comp)
 {
     void* __capability new_ddc = cheri_address_set(manager_ddc, new_comp->base);
-    new_ddc = cheri_bounds_set(new_ddc, new_comp->size + new_comp->scratch_mem_size + new_comp->scratch_mem_stack_size);
+    new_ddc = cheri_bounds_set(new_ddc, new_comp->size + new_comp->scratch_mem_size);
     // TODO bounds double-check
     new_comp->ddc = new_ddc;
 }
 
 void
 comp_add_intercept(struct Compartment* new_comp, uintptr_t intercept_target, struct func_intercept intercept_data)
 {
+    // TODO check whether negative values break anything in all these generated functions
     printf("Found `%s` func at %p.\n", intercept_data.func_name, (void*) intercept_target);
-    int offset = (intercept_data.redirect_func - intercept_target) / 4;
-    assert(offset < (1 << 27));
-    offset &= (1 << 26) - 1;
-    const int arm_b_instr_mask = 0b101 << 26;
-    uint32_t instr_binary = arm_b_instr_mask | offset;
-    int* init_addr_ptr = (void*) intercept_target;
-    struct intercept_patch new_patch = { init_addr_ptr, instr_binary };
+    int32_t new_instrs[INTERCEPT_INSTR_COUNT];
+    size_t new_instr_idx = 0;
+    uintptr_t comp_manager_cap_addr = new_comp->manager_caps + new_comp->active_manager_caps_count * sizeof(void* __capability); // TODO
+
+    const int32_t arm_function_target_register = 0b01010; // use `x10` for now
+    const int32_t arm_transition_target_register = 0b01011; // use `x11` for now
+
+    // movz x0, $target_fn_addr:lo16
+    // movk x0, $target_fn_addr:hi16
+    /*assert(intercept_target < (1 << 32));*/
+    const int32_t arm_movz_instr_mask = 0b11010010100 << 21;
+    const int32_t arm_movk_instr_mask = 0b11110010101 << 21;
+    const ptraddr_t target_address_lo16 = (intercept_data.redirect_func & ((1 << 16) - 1)) << 5;
+    const ptraddr_t target_address_hi16 = (intercept_data.redirect_func >> 16) << 5;
```

Comment:
Does this not handle addresses bigger than 32 bits?

For the general case (often required for patching) you need `movz` and three `movk`s.

Reply:
Yes, it should, but for simplicity, I suggest we leave it as is, and I've added a comment in https://github.com/capablevms/CHERI-ELF-comp/pull/2/commits/575e434efb616b8d21b672763b9d474cbd0368d7. If we integrate VIXL, this will be easier to do anyway.